cassiopeia/solver/HybridSolver.py

```diff
@@ -0,0 +1,332 @@
+
+    def solve(self):
+        """The general hybrid solver routine.
+
+        The hybrid solver proceeds by clustering together cells using the 
+        algorithm stored in the top_solver until a criteria is reached. Once
+        this criteria is reached, the bottom_solver is applied to each
+        subproblem left over from the "greedy" clustering.
+        """
+
+        # call top-down solver until a desired cutoff is reached.
+        _, subproblems = self.apply_top_solver(
+            list(self.unique_character_matrix.index)
+        )
+
+        # multi-threaded bottom solver approach
+        with multiprocessing.Pool(processes=self.threads) as pool:
+
+            results = list(
+                tqdm(
+                    pool.starmap(self.apply_bottom_solver, [(subproblem[0], subproblem[1]) for subproblem in subproblems]),
+                    total=len(subproblems)
+                )
+            )
+
+        for result in results:
+
+            subproblem_tree, subproblem_root = result[0], result[1]
+
+            # check that the only overlapping name is the root, else 
+            # add a new name so that we don't get edges across the tree
+            existing_nodes = [n for n in self.tree]
+            unique_iter = root = (
+                len(self.tree.nodes)
+                - self.unique_character_matrix.shape[0]
+                + self.character_matrix.shape[0]
+            )
+
+            mapping = {}
+            for n in subproblem_tree:
+                if n in existing_nodes and n != subproblem_root:
+                    mapping[n] = unique_iter
+                    unique_iter += 1
+
+            subproblem_tree = nx.relabel_nodes(subproblem_tree, mapping)
+
+            self.tree = nx.compose(self.tree, subproblem_tree)
+
+        # Collapse 0-mutation edges and append duplicate samples
+        self.tree = solver_utilities.collapse_tree(
+            self.tree, True, self.unique_character_matrix, self.missing_char
+        )
+        self.tree = self.append_sample_names(self.tree)
+
+    def apply_top_solver(
+        self, samples: List[int], root: Optional[int] = None
+    ) -> Tuple[int, List[Tuple[int, List[int]]]]:
+        """Applies the top solver to samples.
+
+        A private helper method for applying the top solver to the samples
+        until a criteria is hit. Subproblems and the root ID are returned.
+
+        Args:
+            root: Node ID of the root in the subtree containing the samples.
+            samples: Samples in the subclade.
+
+        Returns:
+            The ID of the node serving as the root of the tree containing the 
+                samples, and a list of subproblems in the form
+                [subtree-root, subtree-samples].
+        """
+
+        mutation_frequencies = self.top_solver.compute_mutation_frequencies(
+            samples
+        )
+
+        clades = list(self.top_solver.perform_split(mutation_frequencies, samples))
```

Comment:
It appears that this change is yet to be merged into `refactor`, so this would actually end up breaking the code in refactor.

When this slight API change is made in `CassiopeiaGreedy.solve()`, we'll see need to change this here as well.